wordpress-mobile/WordPress-Android
# Zendesk WIP: My Tickets & Ticket creation

This is the first PR for Zendesk integration which targets a WIP branch `issue/zendesk-master`. Here is a list of things this PR handles:

* Adds Zendesk dependency
* Sets up Zendesk using credentials from `gradle.properties` (ping me for credentials)
* Adds a lot of helper methods for Zendesk. I decided to go with the Kotlin file with `JvmName` field set, so it'll work like our other `Utils` classes.
* Ticket creation: This has a lot of elements to it, several tags, custom fields etc. I wouldn't recommend comparing it to the existing Helpshift implementation, we are basically writing it from scratch and following iOS implementation's lead on it. (ping me for p2 posts if you're interested in the discussion)
* Zendesk anonymous identity creation for WordPress.com accounts. (self-hosted to be implemented) **NOTE:** Zendesk doesn't allow agent emails as identities, you'll need to use a test account if you have an account in Zendesk with your regular email.
* Support for logging through `SiteModelExtensions`. Instead of using our current logs, we decided to match the iOS one. We might end up improving both of them (I already have an item in the checklist for it)
* Calculating free space in the device: I am using a modified stackoverflow solution for it.
* A few todos - I am adding everything I can think of to our master issue #7737, but I haven't removed the todos I have added from before. Hopefully those will provide some context during review.

This is it for now. I think this covers the fundamentals of the integration and as you can see from #7737 there is a lot more work to come. I was about to remove Helpshift stuff, but that was going to result in a big changeset and I thought it'd be much easier to review that as a standalone thing.

To test:
* **Make sure!!!** to login with an email that doesn't have a Zendesk account attached to it
* Go to Me -> Help & Support -> Contact us and create a ticket. Check the ticket information in Zendesk (tags, logs, fields etc)
* Go to Me -> Help & Support -> My tickets to check that your ticket is there (if you logout and log back in with any other email address, your tickets will be gone! - each time the identity changes, the tickets the user sees will change even if you use an account you have a ticket with, that's how Zendesk handles things)

`ZendeskHelper` file is a work in progress, I am hoping to improve it a bit more maybe while writing tests for it. I am not sure how well it'll work with a singleton, but I'll check it out if I have time.

Finally, an important implementation note is that I opted to re-configure the Zendesk each time a ticket is created or `My Tickets` is opened. Their `singleton` usage is questionable at best. We also have a lot of authentication states and site/log changes. I think this is the best way to ensure that we always use the  most recent information.
